pkg/controllers/machinemigration/machine_migration_controller.go

```diff
@@ -0,0 +1,492 @@
+	}
+
+	return false, nil
+}
+
+// applyStatusAuthoritativeAPIWithPatch updates the resource status.authoritativeAPI using a server-side apply patch.
+func (r *MachineMigrationReconciler) applyStatusAuthoritativeAPIWithPatch(ctx context.Context, m *machinev1beta1.Machine, authority machinev1beta1.MachineAuthority) error {
+	logger := log.FromContext(ctx)
+	logger.Info(fmt.Sprintf("Setting AuthoritativeAPI status to %q for resource", authority))
+
+	ac := machinev1applyconfigs.Machine(m.Name, m.Namespace).
+		WithStatus(machinev1applyconfigs.MachineStatus().WithAuthoritativeAPI(authority))
+
+	if err := r.Status().Patch(ctx, m, util.ApplyConfigPatch(ac), client.ForceOwnership, client.FieldOwner(controllerName+"-AuthoritativeAPI")); err != nil {
+		return fmt.Errorf("failed to patch Machine API machine status with authoritativeAPI %q: %w", authority, err)
+	}
+
+	return nil
+}
+
+// setNewAuthoritativeAPIAndResetSynchronized updates the Machine status to the new authoritativeAPI,
+// resets the synchronized generation, and removes the Synchronized condition.
+func (r *MachineMigrationReconciler) setNewAuthoritativeAPIAndResetSynchronized(ctx context.Context, m *machinev1beta1.Machine, authority machinev1beta1.MachineAuthority) error {
+	var newConditions []*machinev1applyconfigs.ConditionApplyConfiguration
+
+	for _, cond := range m.Status.Conditions {
+		if cond.Type != consts.SynchronizedCondition {
+			newConditions = append(newConditions, &machinev1applyconfigs.ConditionApplyConfiguration{
+				LastTransitionTime: &cond.LastTransitionTime,
+				Message:            &cond.Message,
+				Reason:             &cond.Reason,
+				Status:             &cond.Status,
+				Severity:           &cond.Severity,
+				Type:               &cond.Type,
+			})
+		}
+	}
+
+	statusAc := machinev1applyconfigs.MachineStatus().
+		WithAuthoritativeAPI(authority).
+		WithConditions(newConditions...).
+		WithSynchronizedGeneration(0)
+
+	ac := machinev1applyconfigs.Machine(m.Name, m.Namespace).WithStatus(statusAc)
+
+	if err := r.Status().Patch(ctx, m, util.ApplyConfigPatch(ac), client.ForceOwnership, client.FieldOwner(controllerName+"-AuthoritativeAPI")); err != nil {
+		return fmt.Errorf("failed to patch Machine API machine status with authoritativeAPI %q: %w", authority, err)
+	}
+
+	return nil
+}
+
+// ensureFinalizerOnNewAuthoritativeResource adds a finalizer to the resource if required.
+// If the finalizer already exists, this function should be a no-op.
+// If the finalizer is added, the function will return true so that the reconciler can requeue the object.
+// Adding the finalizer in a separate reconcile ensures that spec updates are separate from status updates.
+func (r *MachineMigrationReconciler) ensureFinalizerOnNewAuthoritativeResource(ctx context.Context, m *machinev1beta1.Machine) (bool, error) {
+	if m.Spec.AuthoritativeAPI != machinev1beta1.MachineAuthorityClusterAPI {
+		return false, nil
+	}
```

Comment:
Why not ensure the MAPI finalizer when going the other way?

We probably just want to make sure the finalizers are on both sides at all times